Framework/NexusGeometry/src/NexusGeometrySave.cpp

```diff
@@ -0,0 +1,805 @@
+ * into std::vector<double> vector. Used by saveInstrument to write array-type
+ * datasets to file.
+ *
+ * @param data : Mantid::Kernel::V3D vector containing data values
+ * @return std::vector<double> vector containing data values in
+ * Mantid::Kernel::V3D format.
+ */
+inline std::vector<double> toStdVector(const V3D &data) {
+  std::vector<double> stdVector;
+  stdVector.reserve(3);
+  stdVector.push_back(data.X());
+  stdVector.push_back(data.Y());
+  stdVector.push_back(data.Z());
+  return stdVector;
+}
+
+/*
+ * Function toStdVector (Overloaded). Store data in Eigen::Vector3d vector into
+ * std::vector<double> vector. Used by saveInstrument to write array-type
+ * datasets to file.
+ *
+ * @param data : Eigen::Vector3d vector containing data values
+ * @return std::vector<double> vector containing data values in Eigen::Vector3d
+ * format
+ */
+inline std::vector<double> toStdVector(const Eigen::Vector3d &data) {
+  return toStdVector(Kernel::toV3D(data));
+}
+
+/*
+ * Function: isApproxZero. returns true if all values in an variable-sized
+ * std-vector container evaluate to zero with a given level of precision. Used
+ * by SaveInstrument methods to determine whether or not to write a dataset to
+ * file.
+ *
+ * @param data : std::vector<T> data
+ * @param precision : double precision specifier
+ * @return true if all elements are approx zero, else false.
+ */
+inline bool isApproxZero(const std::vector<double> &data,
+                         const double &precision) {
+
+  return std::all_of(data.begin(), data.end(),
+                     [&precision](const double &element) {
+                       return std::abs(element) < precision;
+                     });
+}
+
+// overload. return true if vector is approx to zero
+inline bool isApproxZero(const Eigen::Vector3d &data, const double &precision) {
+  return data.isApprox(Eigen::Vector3d(0, 0, 0), precision);
+}
+
+// overload. returns true is angle is approx to zero
+inline bool isApproxZero(const Eigen::Quaterniond &data,
+                         const double &precision) {
+  return data.isApprox(Eigen::Quaterniond(1, 0, 0, 0), precision);
+}
+
+/*
+ * Function: strTypeOfSize
+ * Produces the HDF StrType of size equal to that of the
+ * input string.
+ *
+ * @param str : std::string
+ * @return string datatype of size = length of input string
+ */
+inline H5::StrType strTypeOfSize(const std::string &str) {
+  H5::StrType stringType(H5::PredType::C_S1, str.size());
+  return stringType;
+}
+
+/*
+ * Function: writeStrDataset
+ * writes a StrType HDF dataset and dataset value to a HDF group.
+ *
+ * @param grp : HDF group object.
+ * @param attrname : attribute name.
+ * @param attrVal : string attribute value to be stored in attribute.
+ */
+inline void writeStrDataset(H5::Group &grp, const std::string &dSetName,
+                            const std::string &dSetVal,
+                            H5::DataSpace dataSpace = SCALAR) {
```

Comment:
I think you can pass `dataSpace` as const ref to `writeStrDataset` and both `writeStrAttribute` overloads.

Reply:
done